Add E2E and Unit (PHP and JS) tests for Filter by Stock block.

opr opened the pull request:
<!-- Start by describing the changes made in this Pull Request, and the reason for such changes. -->
This PR introduces E2E, JS unit, and PHP unit tests for the Filter by Stock block.

Besides the code changes to test files, there is one other important change to actual code: I have modified the way that the options for the block are derived and at what point. Originally these options were fetched outside of the component body. Because of this the code was executed as soon as the file was imported. This was not good for testing because we couldn't mock the values returned by `getSetting`. That is because the mocks were set up in the tests after importing the file. The changes made in `assets/js/blocks/stock-filter/block.js` move the options into the component ~~and add them to a `useMemo` with no deps, so this should only be calculated once.~~. 

~~I opted to do this rather than using `useState` because the implication of using this is that the values returned by `getSetting` (and therefore `STOCK_STATUS_OPTIONS`) could change over time, when they will not. I welcome any feedback on this approach though.~~

After discussing with @tjcafferkey we decided that these usages of `useMemo` were "over-optimisations". The length of `STOCK_STATUS_OPTIONS` will _usually_ be 3, so iterating over it is cheap, and not worth memoizing given the complexity this adds to the component.

Because the value of `initialOptions` is used as a dependency to the `useEffect` further down this file, setting these values as internal component state seemed the best way forward to stop this effect running every render.

<!-- Reference any related issues or PRs here -->
Fixes #4661

<!-- Don't forget to update the title with something descriptive. -->
<!-- If your pull request implements a feature flag, make sure you update [this doc](../docs/blocks/features-and-blocks-behind-a-flag.md) -->
### How to test the changes in this Pull Request:

1. Run `npm run test:php` `npm run test` `npm run test:e2e` and verify they all pass.
2. Create a page, add the `Filter by Stock`, `Filter by Price`, `Active filters`, and `All Products` blocks.
3. View the page and ensure the `Filter by Stock` block is working correctly. Verify the counts of each stock level are accurate.
4. Go to **WooCommerce > Settings > Products > Inventory** toggle `Hide out of stock products from catalogue" to be true.
5. View the page again and ensure `Out of stock` is not shown as a stock level option.
6. Wait for tests to go green in this PR.

<!-- If you can, add the appropriate labels -->

opr committed:
Add back end E2E tests for Filter by Stock

opr committed:
Add unit/snapshot tests for Filter by stock block

opr committed:
Refactor Filter by Stock block to load stock statuses in the block body

This is required to make the code easier to test, since the mocked settings were being applied immediately after the block was imported, which didn't give time to let the `allSettings` override take place in the unit tests.

opr committed:
Add PHP Unit tests for test_custom_stock_counts of ProductQueryFilters

tjcafferkey commented on assets/js/blocks/stock-filter/block.js:
```diff
@@ -48,10 +34,31 @@ const StockStatusFilterBlock = ( {
 	attributes: blockAttributes,
 	isEditor = false,
 } ) => {
+	const hideOutOfStockItems = useMemo(
```
Personally this `useMemo` is an over-optimisation to me as there's no real computation happening.

tjcafferkey commented on assets/js/blocks/stock-filter/block.js:
```diff
@@ -48,10 +34,31 @@ const StockStatusFilterBlock = ( {
+		() => getSetting( 'hideOutOfStockItems', false ),
+		[]
+	);
+	const { outofstock, ...otherStockStatusOptions } = useMemo(
```
Personally this `useMemo` is an over-optimisation to me as there's no real computation happening.

tjcafferkey commented on assets/js/blocks/stock-filter/block.js:
```diff
@@ -48,10 +34,31 @@ const StockStatusFilterBlock = ( {
+		() => getSetting( 'stockStatusOptions', {} ),
+		[]
+	);
+	const STOCK_STATUS_OPTIONS = useMemo( () => {
```
Again, seems like an over-optimisation personally for the same reasons as above.

opr committed:
Remove useMemo from stock filter block

This felt like an overoptimisation, it's very inexpensive to run this every render. The result is less complex code, so it's better this way imo.

tjcafferkey commented on assets/js/blocks/stock-filter/block.js:
```diff
@@ -48,10 +34,31 @@ const StockStatusFilterBlock = ( {
 		blockAttributes.isPreview ? previewOptions : []
 	);
+	// Filter added to handle if there are slugs without a corresponding name defined.
+	const initialOptions = useMemo( () => {
```
Unless there are a lot of values within `STOCK_STATUS_OPTIONS` it doesn't seem worth using `useMemo`. Just going to quote Kent C Dodds:

```
Performance optimizations are not free. They ALWAYS come with a cost but do NOT always come with a benefit to offset that cost.

Therefore, optimize responsibly.
```
- [Source](https://kentcdodds.com/blog/usememo-and-usecallback)

opr replied:
Under normal circumstances this will have a length of 3, but merchants _can_ add more stock statuses. After thinking about it, the complexity this brings to the component coupled with the overheads of `useMemo` this don't seem like a worthwhile optimisation in my opinion.

opr committed:
Move initial stock options and the values they are derived from into useState

opr committed:
Add class name to block in editor

ralucaStan commented:
While testing this PR I've come across a bug I've described in https://github.com/woocommerce/woocommerce-gutenberg-products-block/issues/4741
